### DISCUSSION TITLE
Catalyst - Multi Currency Integration

### DISCUSSION THREAD

**Discussion by kathir-arizon** [bigcommerce/catalyst#1514] [General]
Hi Team,

Do we have multi-currency support in the catalyst?

**Comment by chanceaclark** [comment]
Hey @kathir-arizon, we haven't built a currency selector in Catalyst, but we do have the necessary API's to support it. If you want to build it yourself, here's the rough steps I would take:

1. Build a currency selector with the `site.currencies` node.
2. Save selected currency in a cookie of your choice.
3. Use the cookie value when in queries that have a `currencyCode` field.

**Reply by kathir-arizon** [reply]
@chanceaclark  Thanks for your reply. I will look into it.

**Reply by kathir-arizon** [reply]
@chanceaclark, I have integrated steps 1 and 2 but was struck in step 3 when retrieving the cookie and assigning it to the pricingfragment & priceSchema files.

Kindly let me know how to change the value below for currencyCode like CAD to USD or Other Currencies on this page. I tried to retrieve the cookie but nothing is working in this file.

Kindly guide me through some steps here.

File path: 
1) core\client\fragments\pricing.ts
2) core\app\[locale]\(default)\product\[slug]\_components\product-schema.tsx

code:
prices(currencyCode: CAD) {

Dynamic code to retrieve the variable and pass it to the graphql:

![image](https://github.com/user-attachments/assets/75991647-e744-4a5c-9d3a-a4a39238ced3)

Static value of the currencyCode (Working but we have to make it dynamic)
![image](https://github.com/user-attachments/assets/388c5e07-6255-4ada-9eb7-2b6d7412b017)

**Reply by chanceaclark** [reply]
You'll need to add the variable to the query in `core/app/[locale]/(default)/product/[slug]/page-data.ts` like so:
```gql
query ProductPageQuery(
  $entityId: Int!
  $optionValueIds: [OptionValueId!]
  $useDefaultOptionSelections: Boolean
  $currencyCode: currencyCode
) { ... }
```

Then you can do something like this to consume the currencyCode:
```ts
  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  const currencyCode = (await cookies()).get('currencyCode')?.value as CurrencyCode | undefined;

  const product = await getProduct({
    entityId: productId,
    optionValueIds,
    useDefaultOptionSelections: true,
    currencyCode,
  });
```


P.S. I had to add this type in `page-data.ts` for the casting to work:
```ts
export type CurrencyCode = Variables['currencyCode'];
```

**Reply by kathir-arizon** [reply]
@chanceaclark Thanks for your response. It has a lot of dependencies and am fixing it one by one. It's working now :)

**Reply by kathir-arizon** [reply]
@chanceaclark I found one issue when we changed the currency setting the cookie and making the router.refresh it is converting the price based on the currency but when we navigate to the PLP to PDP it still shows the site's default currency bcoz the cookie is loading as undefined.

Do you know if this on-page navigation cookie is coming as undefined?

**Reply by chanceaclark** [reply]
I bet it's because the guest shopper experience is a static page rather than dynamic. If you log in, it should dynamic. I would guess the alternatives here are:
1. Make the page fully dynamic *Not Recommended*
2. Extract the price fetching into a server component in a suspense boundary. `RelatedProducts` is a good example of this. You can check out the Next.js documentation on PPR to give you a rough idea of this approach (without PPR actually working).

**Reply by kathir-arizon** [reply]
@chanceaclark  Thanks for the clarification. I will look into it.

**Reply by kathir-arizon** [reply]
@chanceaclark I have changed the PDP pricing logic to suspense and still see the cookie undefined issue.  We are having the same price issue with the RelatedProducts.

When I enable the PPR and it throws an error to install the nextjs latest canary. Kindly recommend which version is suitable for the canary to enable the PPR.

The current Next Js version used in Catalyst is "14.2.13". If we switch to the latest canary it will upgrade to 15.x.x Is it fine to go with next 15?

Error: The experimental.ppr preview feature can only be enabled when using the latest canary version of Next.js. See more info here: https://nextjs.org/docs/messages/ppr-preview

Kindly provide your recommendation.

**Reply by kathir-arizon** [reply]
@chanceaclark Could you please let me know how to disable cache for guest flow? I have installed the next js 15 canary and used the suspense boundary but the server component returns the cookies as undefined. I have managed to retrieve the data using the useEffect on the client component but In PLP, prices are updating one by one product and it doesn't look good.

I have cross-verified your commit and am still getting the same issue. https://github.com/bigcommerce/catalyst/commit/5555eda622a05bc53490e919c9ecd1bea3667192#diff-791b48067710e5f48f45a218435f5a7ab9b4bec30b7d75983b473dfd95fe633fR114

**Reply by chanceaclark** [reply]
Hey @kathir-arizon for guest shoppers we static page to improve performance for shoppers. You may be running into this specifically. You might want to look into [Next.js Route Segments](https://nextjs.org/docs/app/api-reference/file-conventions/route-segment-config#dynamic) to see how you could tailor these pages more.

However, we have a ticket to add a currency selector in the next couple of weeks or so if you want to wait till we wrap that up?

**Reply by kathir-arizon** [reply]
@chanceaclark  Do we have any PR for that or any branch to check? We will upgrade it once it is live but I have removed the category page level cache to fix the guest customers.

Thanks for your help @chanceaclark

**Comment by kathir-arizon** [comment]
>

**Comment by developersupreme** [comment]
@kathir-arizon Did you successfully implement multi-currency functionality for pricing? We're looking to implement the same and would appreciate it if you could share the steps, if possible.

**Reply by kathir-arizon** [reply]
@developersupreme  We have integrated it.

**Comment by chanceaclark** [comment]
Hey everyone, thank you for waiting patiently! We just opened up a PR for our Currency Selector: https://github.com/bigcommerce/catalyst/pull/1912

The only issue with it right now is that we don't get back the default currency from the API but we are actively working on adding that. Test it out and let us know if you have any feedback.
